target/snitch_cluster/sw/snax/streamer-simd/src/snax-streamer-simd-lib.c

```diff
@@ -0,0 +1,162 @@
+// Copyright 2023 KU Leuven.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Xiaoling Yi <xiaoling.yi@esat.kuleuven.be>
+
+#include "snax-streamer-simd-lib.h"
+
+int32_t gen_csr0_config(uint8_t input_zp_i, uint8_t output_zp_i,
+                        uint8_t shift_i, uint8_t max_int_i) {
+    // encode the configuration into a single 32-bit integer
+    return ((int32_t)max_int_i << 24) | ((int32_t)shift_i << 16) |
+           ((int32_t)output_zp_i << 8) | (int32_t)input_zp_i;
+}
+
+int32_t gen_csr1_config(uint8_t min_int_i, bool double_round_i) {
+    // encode the configuration into a single 32-bit integer
+    return ((uint8_t)double_round_i << 8) | (uint8_t)min_int_i;
+}
+
+int32_t gen_csr2_config(uint32_t multiplier_i) { return multiplier_i; }
+
+void set_streamer_simd_csr(int tempLoop0, int tempLoop1, int tempStride0_in,
+                           int tempStride1_in, int tempStride0_out,
+                           int tempStride1_out, int32_t delta_local_in,
+                           int32_t delta_local_out) {
+    // temporal loop bounds, from innermost to outermost
+    write_csr(960 + 0, tempLoop0);
+    write_csr(960 + 1, tempLoop1);
+
+    // temporal strides for data reader (In)
+    write_csr(960 + 2, tempStride0_in);
+    write_csr(960 + 3, tempStride1_in);
+
+    // temporal strides for data writer (Out)
+    write_csr(960 + 4, tempStride0_out);
+    write_csr(960 + 5, tempStride1_out);
+
+    // fixed spatial strides for data reader (In)
+    write_csr(960 + 6, 4);
+    write_csr(960 + 7, 32);
+
+    // fixed spatial strides for data writer (Out)
+    write_csr(960 + 8, 1);
+    write_csr(960 + 9, 8);
+
+    // base ptr for data reader (In)
+    write_csr(960 + 10, (uint32_t)(delta_local_in + snrt_l1_next()));
+
+    // base ptr for data writer (Out)
+    write_csr(960 + 11, (uint32_t)(delta_local_out + snrt_l1_next()));
+}
+
+void start_streamer_simd() { write_csr(960 + 12, 1); }
+
+void set_simd_csr(uint32_t csr0, uint32_t csr1, uint32_t csr2,
+                  uint32_t temporal_loop_bound) {
+    // set the constants for the SIMD unit
+    write_csr(960 + 14, csr0);
+    write_csr(960 + 15, csr1);
+    write_csr(960 + 16, csr2);
+
+    // set the temporal loop bound
+    write_csr(960 + 17, temporal_loop_bound);
+}
+
+void start_simd() { write_csr(960 + 18, 1); }
+
+void wait_streamer_simd() {
+    write_csr(960 + 18, 0);
+    write_csr(960 + 18, 0);
+    write_csr(960 + 12, 0);
+}
```

Comment:
You are waiting for the same thing twice here!
Could you maybe annotate which is which? e.g. streamer IN, SIMD?